File: src/backtracking/python/all_combinations/src/all_combinations.py
from typing import List


def all_combinations(n: int, k: int) -> List[List[int]]:
    """
    Generate all combinations of k numbers from 0 to n.

    Uses backtracking to find all combinations of size k from the range [0, n].

    Args:
        n: The upper bound of the range (inclusive).
        k: The size of each combination.

    Returns:
        A list of all combinations, where each combination is a list of k integers.
    """

    def _all_combinations(combination: List[int]) -> None:
        if len(combination) == k:
            result.append(combination)
            return

        for i in range(combination[-1] + 1 if combination else 0, n + 1):
            _all_combinations(combination + [i])

    result: List[List[int]] = []
    _all_combinations([])
    return result

File: src/backtracking/cpp/all_combinations/src/all_combinations.cpp
#include "all_combinations.h"
#include <functional>

std::vector<std::vector<int>> all_combinations(int n, int k) {
  std::vector<std::vector<int>> result;

  // Nested function to generate combinations
  std::function<void(std::vector<int>)> _all_combinations;
  _all_combinations = [&](std::vector<int> combination) {
    if (combination.size() == k) {
      result.push_back(combination);
      return;
    }

    for (int i = (combination.empty() ? 0 : combination.back() + 1); i <= n;
         ++i) {
      combination.push_back(i);
      _all_combinations(combination);
      combination.pop_back();
    }
  };

  // Initialize result and call nested function
  _all_combinations({});
  return result;
}
